src/iterator.jl

```diff
@@ -0,0 +1,291 @@
+abstract type VariantIterator end
+abstract type GeneticData end 
+
+mutable struct VCFIterator <: VariantIterator
+    vcffile::AbstractString
+    vcf::VCF.Reader
+    sample_num::Int
+end
+
+function VCFIterator(vcffile::AbstractString; sample_num::Int=VCFTools.nsamples(VCF.Reader(openvcf(vcffile, "r"))))
+    vcf = VCF.Reader(openvcf(vcffile, "r")) 
+    return VCFIterator(vcffile, vcf, sample_num) #then returns an iostream 
+end
+
+mutable struct VCFData <: GeneticData # feed VCFData VCFIterator function output output of openvcf function 
+    file_name::AbstractString
+end
+# placeholder because other genotype files need that 
+
+mutable struct VCFRow <: GeneticVariantBase.Variant
+    CHROM::String
+    POS::Int64
+    ID::Vector{String}
+    REF::String
+    ALT::Vector{String}
+    QUAL::Float64
+    GENOTYPE::Vector{Union{Float64, Missing}}
+    DOSAGES::Union{Nothing, Vector{Union{Missing, Float64}}}
+    INDEX::Int64
+end
+
+# for dosages there is a key DS 
+# copy_gt! and copy_ds!
+# for genotypes taking values from GT
+
+# return VCFRow item from the iterate function 
+
+@inline function Base.eltype(::Type{<:VariantIterator}) # vector of string 
+    VCFRow
+end
+
+function Base.iterate(itr::VCFIterator, state=1)
+    rows = nrecords(itr.vcffile)
+
+    if state <= 0 || state > rows
+        return nothing
+    else
+        reader = VCF.Reader(openvcf(itr.vcffile, "r")) 
+        count = 0
+
+        for record in reader
```

Comment:
This for loop should be removed, and each call to this function should be O(N) time, where N is the number of samples. It should be constant in terms of number of genetic variants.